baseapp/baseapp.go

```diff
@@ -81,6 +75,7 @@ type BaseApp struct {
 	postHandler sdk.PostHandler // post handler, optional, e.g. for tips
 
 	initChainer          sdk.InitChainer                // ABCI InitChain handler
+	preBlocker           sdk.PreBlocker                 // logic to run before BeginBlocker
```

Comment:
I try to merge #16898 into this PR, but need at least 4 [changes](https://github.com/mmsqe/cosmos-sdk/compare/3ffb4e9cbf6b3b5ee84a22b4bf0a745658be24bb...126109182353903a90328278b4124c0948747353) to achieve same goal:
1. add return `ResponsePreBlock` in `PreFinalizeBlockHook` to migrate related check to avoid refreshing ConsensusParams of ctx every time
2. remove preBlocker property
3. replace `PreBlocker` with `PreFinalizeBlockHook`
4. replace `SetPreBlock` with `SetPreFinalizeBlockHook` of `PreFinalizeBlockHook`

Reply:
`preFinalizeBlockHook` can be re-implemented as a `PreBlocker` in a module?

Reply:
This is what we talked about recently, yes @yihuang. Maybe we pursue that isntead?

Reply:
Yeah, I guess the plan is:
- merge this PR as is and backport to 0.50
- reimplement preFinalizeBlockHook as PreBlocker in 0.51

Or can we reimplement preFinalizeBlockHook in 0.50 directly?

Reply:
> Or can we reimplement preFinalizeBlockHook in 0.50 directly?

This is the better path IMO. I believe we should do that.

cc @julienrbrt

Reply:
May I ask if we need reimplement `preFinalizeBlockHook` as `PreBlocker` in another PR or in this PR?

Reply:
I personally would love to see the big picture with how it would look like, but I think a second PR based on this one is better for reviewing.

Reply:
Sure, I draft the 2nd [PR](https://github.com/mmsqe/cosmos-sdk/pull/2/files) and can adjust rest doc change later

Reply:
Imo we should have a pre and post proposal and prepare. and this is kinda silly.

Reply:
This PR is just to resolve the consensus param refreshing issue with a pattern to maintain pre-beginBlock logic in BaseApp, not sure if we really need your suggested hooks.

Reply:
> Imo we should have a pre and post proposal and prepare. and this is kinda silly.

What I like is the naming is Comet agnostic, which is handy when there will be more consensus supported.

Reply:
Why should there be hooks on `Prepare` and `Process` proposal @itsdevbear? The app has complete control over those -- it would be completely redundant.

Reply:
resolving this to keep the scope small. @itsdevbear lets open an issue to hold discussion on your design requests